src/Middleware/BasicNetworkResolver.php

```diff
@@ -0,0 +1,141 @@
+        'http' => ['http'],
+        'https' => ['https', 'on'],
+    ];
+
+    private $protocolHeaders = [];
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $newScheme = null;
+        foreach ($this->protocolHeaders as $header => $data) {
+            if (!$request->hasHeader($header)) {
+                continue;
+            }
+            $headerValues = $request->getHeader($header);
+            if (is_callable($data)) {
+                $newScheme = call_user_func($data, $headerValues, $header, $request);
+                if ($newScheme === null) {
+                    continue;
+                } elseif (!is_string($newScheme)) {
+                    throw new \RuntimeException('The scheme is neither string nor null!');
+                } elseif (strlen($newScheme) === 0) {
+                    throw new \RuntimeException('The scheme cannot be an empty string!');
+                }
+                break;
+            }
+            $headerValue = strtolower($headerValues[0]);
+            foreach ($data as $protocol => $acceptedValues) {
+                if (!in_array($headerValue, $acceptedValues)) {
+                    continue;
+                }
+                $newScheme = $protocol;
+                break 2;
+            }
+        }
+        $uri = $request->getUri();
+        if ($newScheme !== null && $newScheme !== $uri->getScheme()) {
+            $request = $request->withUri($uri->withScheme($newScheme));
+        }
+        return $handler->handle($request);
+    }
+
+    /**
+     * With header to check for determining whether the connection is made via HTTP or HTTPS (or any protocol).
+     *
+     * The match of header names and values is case-insensitive.
+     * It's not advisable to put insecure/untrusted headers here.
+     *
+     * Accepted types of values:
+     * - NULL (default): {{DEFAULT_PROTOCOL_AND_ACCEPTABLE_VALUES}}
+     * - callable: custom function for getting the protocol
+     * ```php
+     * ->withProtocolHeader('x-forwarded-proto', function(array $values, string $header, ServerRequestInterface $request) {
+     *   return $values[0] === 'https' ? 'https' : 'http';
+     *   return null;     // If it doesn't make sense.
+     * });
+     * ```
+     * - array: The array keys are protocol string and the array value is a list of header values that indicate the protocol.
+     * ```php
+     * ->withProtocolHeader('x-forwarded-proto', [
+     *   'http' => ['http'],
+     *   'https' => ['https']
+     * ]);
+     * ```
+     *
+     * @param callable|array|null $protocolAndAcceptedValues
+     * @return static
+     * @see DEFAULT_PROTOCOL_AND_ACCEPTABLE_VALUES
+     */
+    public function withProtocolHeader(string $header, $protocolAndAcceptedValues = null)
+    {
+        $new = clone $this;
+        $header = strtolower($header);
+        if ($protocolAndAcceptedValues === null) {
```

Comment:
I'd like to get rid of the if/elseif construction to enhance readability.
Would it be possible to return `$new` right away or throw ?

Reply:
Yes, modified.